### PR TITLE
Add simple RTSP helper plugin

### PR DESCRIPTION
**Pull request by QuentinC**
Hello,

This is a simple plugin that I have made to allow some RTSP streams to work behind a NAT firewall.
The main reason is the TV service offered by m provider (French internet provider named Free), that provides RTSP streams to watch TV on the router LAN.

This stream won't work because the server ports used are refusing UDP packets, with an ICMP Port Unreachable return. This causes the firewall to cancel the NAT inbound port, and then blocks the video stream.

The plugin is there to intercept RTSP TCP connections, and watching for "client_port" option. It then creates inbound redirections of these ports, allowing the video to pass. This port remains opened as long as the TCP connection stays opened, and is checked against access rules (for both IP and ports). 

This plugin acts as a TCP proxy server for the TCP RTSP stream. It is fully configurable from the web UI.

I can provide PCAP files showing the "fixed" issue.

Thanks !

### CODE REVIEW

File: net/rtsphelper/src/opnsense/scripts/net/rtsphelper/rtsphelper.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    signal.signal(signal.SIGTERM, handle_exit_signal)
+    try:
+        while 1:
+            time.sleep(delay)
+            for server in servers:
+                server.main_loop()
+    except KeyboardInterrupt:
+        print("Ctrl C - Stopping server")
+        handle_exit()
+sys.exit(1)
```
Comment (fabianfrz) [inline net/rtsphelper/src/opnsense/scripts/net/rtsphelper/rtsphelper.py]: add newline

**Comment by fabianfrz** [comment]
@QuentinC for new plugins, the legacy frontend will not be accepted. They must be developed using the MVC components. You can look at the [os-helloworld plugin](https://github.com/opnsense/plugins/tree/master/devel/helloworld) to get a template. (your main work here will be that you will have to split the view and the controller code)

pfctl calls need root privileges so you need to run them by configd - so you need to define actions for them.

File: net/rtsphelper/Makefile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+PLUGIN_NAME=		rtsphelper
+PLUGIN_VERSION=		1.0
+#PLUGIN_DEPENDS=
```
Comment (fabianfrz) [inline net/rtsphelper/Makefile]: remove it

File: net/rtsphelper/src/etc/inc/plugins.inc.d/rtsphelper.inc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    mwexec_bg('/usr/local/bin/python2.7 /usr/local/opnsense/scripts/net/rtsphelper/rtsphelper.py');
+}
+
+function rtsphelper_stop()
+{
+    killbypid('/var/run/rtsphelper.pid', 'TERM', true);
+    mwexec('/sbin/pfctl -artsphelper -Fr 2>&1 >/dev/null');
+    mwexec('/sbin/pfctl -artsphelper -Fn 2>&1 >/dev/null');
```
Comment (fabianfrz) [inline net/rtsphelper/src/etc/inc/plugins.inc.d/rtsphelper.inc]: call it via configd - this needs root

File: net/rtsphelper/src/www/services_rtsphelper.php [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        filter_configure();
+        header(url_safe('Location: /services_rtsphelper.php'));
+        exit;
+    }
+}
+
+
+$service_hook = 'rtsphelper';
+legacy_html_escape_form_data($pconfig);
+include("head.inc");
```
Comment (fabianfrz) [inline net/rtsphelper/src/www/services_rtsphelper.php]: from here split into controller and view

File: net/rtsphelper/src/www/services_rtsphelper.php [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                  <thead>
+                    <tr>
+                      <th colspan="2"><?=gettext("User specified permissions");?></th>
+                    </tr>
+                  </thead>
+                  <tbody>
+<?php foreach (rtsphelper_permuser_list() as $i => $permuser): ?>
+                    <tr>
+<?php if ($i == 1): ?>
+                      <td style="width:22%"><a id="help_for_permuser" href="#" class="showhelp"><i class="fa fa-info-circle"></i></a> <?=gettext('Entry') . ' ' . $i ?></td>
```
Comment (fabianfrz) [inline net/rtsphelper/src/www/services_rtsphelper.php]: should be a template using sprintf

File: net/rtsphelper/src/www/status_rtsphelper.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+$rdr_entries = array();
+exec("/sbin/pfctl -artsphelper -sn", $rdr_entries, $pf_ret);
+
+$service_hook = 'rtsphelper';
+include("head.inc");
+?>
+<body>
+<?php include("fbegin.inc"); ?>
```
Comment (fabianfrz) [inline net/rtsphelper/src/www/status_rtsphelper.php]: split here

File: net/rtsphelper/src/www/status_rtsphelper.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    if (!empty($_POST['clear'])) {
+        rtsphelper_stop();
+        rtsphelper_start();
+        header(url_safe('Location: /status_rtsphelper.php'));
+        exit;
+    }
+}
+
+$rdr_entries = array();
+exec("/sbin/pfctl -artsphelper -sn", $rdr_entries, $pf_ret);
```
Comment (fabianfrz) [inline net/rtsphelper/src/www/status_rtsphelper.php]: must use configd

**Comment by QuentinC** [comment]
Sadly I won't have the time/skills to do the migration to the MVC framework... This plugin is mainly a copy of the UPnP plugin, so if UPnP gets updated I will update this one too.

**Comment by QuentinC** [comment]
For reference, here is the packaged devel file. 
It can be installed after unzipping with: 
`pkg add os-rtsphelper-devel-1.0.txz`

[os-rtsphelper-devel-1.0.txz.zip](https://github.com/opnsense/plugins/files/4094406/os-rtsphelper-devel-1.0.txz.zip)

**Comment by QuentinC** [comment]
Simple upgrade for the package that supports new versions of OPNsense without Python 2.7.
Upgrade with :
`pgk upgrade os-rtsphelper-devel-1.1.txz`

[os-rtsphelper-devel-1.1.txz.zip](https://github.com/opnsense/plugins/files/4216842/os-rtsphelper-devel-1.1.txz.zip)

**Comment by lecossais** [comment]
Thanks for this plugin but not function totally, i see in status my 2 port int (plugin rstp helper)
but no see Automatically forward. 
is there another configuration to do in opnsense?

**Comment by QuentinC** [comment]
> 
> 
> Thanks for this plugin but not function totally, i see in status my 2 port int (plugin rstp helper)

Hello,


> but no see Automatically forward.

The rules created with this plugin wont appear in the firewall rules if this was what you expected

> is there another configuration to do in opnsense?

If you see the port forward in status page then the helper has successfully seen and enabled the connection.
The helper only creates port forwarding in a special table on pf for inbound UDP traffic. There is no need to add any rule in the firewall.

**Comment by lecossais** [comment]
Thanks my firewall block on my server. Thank you for this workEnvoyé depuis mon téléphone Huawei-------- Message original --------De : QuentinC <notifications@github.com>Date : mer. 26 févr. 2020 à 21:58À : opnsense/plugins <plugins@noreply.github.com>Cc : lecossais <brault.seb@gmail.com>, Comment <comment@noreply.github.com>Objet : Re: [opnsense/plugins] Add simple RTSP helper plugin (#1312)
Thanks for this plugin but not function totally, i see in status my 2 port int (plugin rstp helper)

Hello,

but no see Automatically forward.

The rules created with this plugin wont appear in the firewall rules if this was what you expected

is there another configuration to do in opnsense?

If you see the port forward in status page then the helper has successfully seen and enabled the connection.
The helper only creates port forwarding in a special table on pf for inbound UDP traffic. There is no need to add any rule in the firewall.

**Comment by QuentinC** [comment]
Small upgrade to follow changes on miniupnpd (remove services.inc)
Here is the package: [os-rtsphelper-devel-1.2.txz.zip](https://github.com/opnsense/plugins/files/5198352/os-rtsphelper-devel-1.2.txz.zip)

Upgrade with : `pgk upgrade os-rtsphelper-devel-1.1.txz`

**Comment by pkx232c** [comment]
Hi Quentin,

I have the same issue with Free and I'm trying to get your plugin working with PfSense.
So far I'm trying to run the Python script manually and to figure out what to put in the config file.

This is my guess, is that correct ?
ext_ifname=eth0 -> name of the physical interface for wan
forward=212.27.38.253:554 -> IP and port to forward RTSP to
allow=192.168.1.1/24 30000-65535 -> Subnet allowed to be redirected to

I'm far from Python expert needless to say..

Thanks for the good job :-)

**Comment by QuentinC** [comment]
> 
> 
> Hi Quentin,
> 
> I have the same issue with Free and I'm trying to get your plugin working with PfSense.
> So far I'm trying to run the Python script manually and to figure out what to put in the config file.
> 
> This is my guess, is that correct ?
> ext_ifname=eth0 -> name of the physical interface for wan
> forward=212.27.38.253:554 -> IP and port to forward RTSP to
> allow=192.168.1.1/24 30000-65535 -> Subnet allowed to be redirected to
> 
> I'm far from Python expert needless to say..
> 
> Thanks for the good job :-)

Hello,

This is exactly the settings that I am using at home !

Best regards !

**Comment by pkx232c** [comment]
Thank you Quentin !

I'm getting close to get it working, I need to figure out how to create anchors in the firewall now ;-)

I will keep you informed of course

**Comment by QuentinC** [comment]
In OPNsense this is done here:
[https://github.com/opnsense/plugins/pull/1312/files#diff-dbe29390eac500a2e4a33fc07fd26963a2f4c6471a86524832474d9e0c611b70R16](https://github.com/opnsense/plugins/pull/1312/files#diff-dbe29390eac500a2e4a33fc07fd26963a2f4c6471a86524832474d9e0c611b70R16)

**Comment by elvquant** [comment]
@pkx232c  Looking for this helper on pfSense.

**Comment by pkx232c** [comment]
@elvquant Unfortunately I didn't have much time to work on it since my last message :-(
I was planning to resume work on this in the next weeks, I'll let you know if I make any progress.

Have a nice day ;-)

**Comment by elvquant** [comment]
@pkx232c many thanks in advanced. I talk to pfSense support the last days and they have no solution for a RTSP through its FW. 

Have a nice weekend and happy eastern.

**Comment by fraenki** [comment]
@fichtner As pointed out by @fabianfrz this PR is based on the legacy codebase, so unfortunately I'd say this PR should be closed.

**Comment by fichtner** [comment]
Yes I agree. We are happy to review and GUI an MVC version of this though. A new draft PR is all it takes.


Cheers,
Franco

**Comment by C-Duv** [comment]
@QuentinC have you planned to re-open a PR for your RTSP helper plugin with non-legacy base? Or would it take too much work to re-implement your solution?

**Comment by C-Duv** [comment]
Forgot to add: if the "legacy" issue only concerns the UI/frontend, I can take a look to:

* [`net/rtsphelper/src/opnsense/mvc/app/models/Net/RTSPHelper`](https://github.com/QuentinC/opnsense-plugins/tree/62a9a29aa38bace4622b60d5efd9d1ee40cb84ec/net/rtsphelper/src/opnsense/mvc/app/models/Net/RTSPHelper)
* [`net/rtsphelper/src/www`](https://github.com/QuentinC/opnsense-plugins/tree/62a9a29aa38bace4622b60d5efd9d1ee40cb84ec/net/rtsphelper/src/www)

(to make follow new standard as the `os-helloworld plugin`)

**Comment by QuentinC** [comment]
Hello @C-Duv ,

The issues are:
1. The Frontend/UI:
I haven't been able to make the required changes mostly because of my basic web dev knowledge.
For your information, this plugin is a copy of the net-upnp plugin (https://github.com/opnsense/plugins/tree/master/net/upnp), there are only a few changes on the labels of the UI, and the python script of course.

I've checked right now the net-upnp still uses the legacy UI, if you know how to migrate to MVC feel free to update this PR !

2. The "pfctl" calls, that should be handled by configd.
This shouldn't be too hard to fix, but I haven't looked at it yet

**Comment by fichtner** [comment]
os-upnp is not a good example, because it came via fork from the old code base and it's been static ever since.

**Comment by JGeek00** [comment]
Hi. Any news about this?
I have the same problem as @QuentinC has. My ISP serves RTSP streams for the IPTV service and I don't know how to handle it with OPNSense.

**Comment by ghettopato** [comment]
is there any way to get an updated version of this?

**Comment by jgk-bckp** [comment]
Hi @LirycR 
When could it be available on the OPNSense plugins list to be installed?

**Comment by myde2001** [comment]
Hi @QuentinC @LirycR it appears that it doesn't work in OPNsense 25.1. Does anyone have it working?

**Pull request closed without merging**